### PR TITLE
Reland "Migrate darwin common "framework_shared" target to ARC #37049" 

### PR DESCRIPTION
**Pull request by cyanglaz**
Relands https://github.com/flutter/engine/pull/37219, which was reverted at https://github.com/flutter/engine/pull/37320 due to a benchmark regression. https://github.com/flutter/flutter/issues/114697

There is no change in this PR compares to https://github.com/flutter/engine/pull/37219

The benchmark regression is due to a performance regression caused by ARC, specifically here: https://github.com/flutter/engine/blob/main/shell/platform/darwin/common/framework/Source/FlutterStandardCodec.mm#L458-L464 . 
This regression is unavoidable and been discussed at several places online: https://jimmymandersson.medium.com/understanding-arcs-effect-on-your-apps-performance-377e39076a88

I create two simple sample test cases comparing ARC and ARC in performance and the result shows (from my machine) that the ARC version of the test took 8ms and MRC took 6ms: https://github.com/cyanglaz/MrcVSArc

ARC has many benefits so I think we should still migrate to it. Please advise if there are any to optimize the ARC code in this PR to reduce the regression.

I also have tested the performance if all the "copy"s are reverted to "retain", but the improvement is very minimal. I'd suggest keep them as copy so the code is safer. 

part of https://github.com/flutter/flutter/issues/112232

cc @zanderso @stuartmorgan @gaaclarke for thoughts. 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by stuartmorgan** [comment]
> The benchmark regression is due to a performance regression caused by ARC, specifically here: https://github.com/flutter/engine/blob/main/shell/platform/darwin/common/framework/Source/FlutterStandardCodec.mm#L458-L464 .

I don't have any specific suggestions for optimizing that code under ARC. That said, I'm also not super concerned about the exact perf of that piece of code; anyone passing huge amounts of data over method channels as a `List<Object>` is going to have a bad time regardless, and the regression wasn't huge. (15% isn't ideal of course, but it's not like it was doubling or something.)

> ARC has many benefits so I think we should still migrate to it.

Agreed, I would not recommend holding back ARC migration on this benchmark regression. I would expect a non-trivial reduction in the number of crashes in Flutter applications over time by migrating to ARC, in addition to easier maintenance (which frees eng resources for other tasks--like going back and optimizing critical codepaths), both of which are very significant benefits.

> I also have tested the performance if all the "copy"s are reverted to "retain", but the improvement is very minimal.

That seems expected to me; the primary foundation types all use CoW under the hood, I believe.

**Comment by gaaclarke** [comment]
Here's the loop with ARC:
```s
LBB2_2:                                 ;   in Loop: Header=BB2_4 Depth=1
	ldr	x0, [x24, _OBJC_CLASSLIST_REFERENCES_$_.1@PAGEOFF]
	bl	_objc_msgSend$null
	mov	x29, x29
	bl	_objc_retainAutoreleasedReturnValue
	mov	x23, x0
	mov	x0, x21
	mov	x2, x23
	bl	"_objc_msgSend$addObject:"
	mov	x0, x23
	bl	_objc_release
LBB2_3:                                 ;   in Loop: Header=BB2_4 Depth=1
	mov	x0, x22
	bl	_objc_release
	subs	w20, w20, #1
	b.eq	LBB2_6
LBB2_4:                                 ; =>This Inner Loop Header: Depth=1
	mov	x0, x19
	bl	_objc_msgSend$readValue
	mov	x29, x29
	bl	_objc_retainAutoreleasedReturnValue
	mov	x22, x0
	cbz	x0, LBB2_2
; %bb.5:                                ;   in Loop: Header=BB2_4 Depth=1
	mov	x0, x21
	mov	x2, x22
	bl	"_objc_msgSend$addObject:"
	b	LBB2_3
```

In order to get rid of all the extra retains and releases I used the `__unsafe_unretained` annotation, which should give you the better performance.

```objc
-(NSArray*)readArray {
  UInt32 length = [self readSize];
  NSMutableArray* array = [NSMutableArray arrayWithCapacity:length];
  for (UInt32 i = 0; i < length; i++) {
    __unsafe_unretained id value = [self readValue];
    [array addObject:(value == nil ? [NSNull null] : value)];
  }
  return array;
}
```

```s
LBB2_2:                                 ;   in Loop: Header=BB2_3 Depth=1
	ldr	x0, [x23, _OBJC_CLASSLIST_REFERENCES_$_.1@PAGEOFF]
	bl	_objc_msgSend$null
	mov	x29, x29
	bl	_objc_retainAutoreleasedReturnValue
	mov	x22, x0
	mov	x0, x21
	mov	x2, x22
	bl	"_objc_msgSend$addObject:"
	mov	x0, x22
	bl	_objc_release
	subs	w20, w20, #1
	b.eq	LBB2_5
LBB2_3:                                 ; =>This Inner Loop Header: Depth=1
	mov	x0, x19
	bl	_objc_msgSend$readValue
	mov	x29, x29
	bl	_objc_unsafeClaimAutoreleasedReturnValue
	cbz	x0, LBB2_2
; %bb.4:                                ;   in Loop: Header=BB2_3 Depth=1
	mov	x2, x0
	mov	x0, x21
	bl	"_objc_msgSend$addObject:"
	subs	w20, w20, #1
	b.ne	LBB2_3
```

I didn't benchmark them, but it should be faster.  Although maybe not as fast as MRC because of the `_objc_unsafeClaimAutoreleasedReturnValue` call.

**Comment by gaaclarke** [comment]
To be clear, after investigating the problem I don't feel there is any legitimate reason to accept this loss of performance and this is something we work hard to optimized.  You can either pull out the loop to an MRC file or use the `__unsafe_unretained` annotation in conjunction with enabling ARC.

**Comment by cyanglaz** [comment]
I tried `__unsafe_unretained`, which crashed because ARC released the returned object of `[self readValue]` right away, making `value` pointing to an invalid address. 

Having this method in MRC is OK in this case for the sake of making benchmark test happy. 

However, there are many performance reductions like this as we migrate to ARC (Some might be caught by benchmarks some might not). If we are going to live with a partial MRC embedder for performance reason, we have to define some rules to draw a line between what we migrate and what we don't.

**Comment by gaaclarke** [comment]
> I tried `__unsafe_unretained`, which crashed because ARC released the returned object of `[self readValue]` right away, making `value` pointing to an invalid address.

Probably not worth debugging since those functions like `objc_unsafeClaimAutoreleasedReturnValue` aren't really documented and we have a workaround, but I'd be curious to see if that is the guy that is clearing out the variable.

> we have to define some rules to draw a line between what we migrate and what we don't.

The line here is this is a particularly sensitive benchmark that potential customers evaluate when making decisions about adopting Flutter or not.

**Comment by cyanglaz** [comment]
> The line here is this is a particularly sensitive benchmark that potential customers evaluate when making decisions about adopting Flutter or not.

I guess the rule for the migration for now would be if we break a benchmark, we need to evaluate it case by case and decide if we should have some code remain in MRC.

**Comment by stuartmorgan** [comment]
> The line here is this is a particularly sensitive benchmark that potential customers evaluate when making decisions about adopting Flutter or not.

What's the real-world use case for a huge number of objects in a list over platform channels?

**Comment by gaaclarke** [comment]
> What's the real-world use case for a huge number of objects in a list over platform channels?

Where is the guarantee that it isn't a use-case?  Platform channel performance is of particular importance to customers evaluating Flutter for add-to-app.  That isn't a hypothetical.  I'm happy to discuss privately.  I'll also add that that the list doesn't have to be large, chatty smaller payloads would experience the same overhead.  Decoding messages happens on the ui thread where we are talking about scales of 16ms to maintain smooth animations.

**Comment by stuartmorgan** [comment]
> Where is the guarantee that it isn't a use-case?

Nowhere. My position is based on the widely held industry view that premature optimization is generally a bad thing, so the bar for making code more complex and more difficult to maintain is not "do it unless you can prove that performance will never be a problem", but rather "only do it if there's concrete evidence that it is".

> I'll also add that that the list doesn't have to be large, chatty smaller payloads would experience the same overhead.

What is the impact (either in ms or percentage of overall message time) of just this retain/release in a list with, say, 10 objects? Maybe I'm misunderstanding the benchmark; I thought it was visible here only because of having a very high element count.

> Platform channel performance is of particular importance to customers evaluating Flutter for add-to-app. That isn't a hypothetical.

Does this benchmark reflect real-world-scenario performance though? Or is it accidentally and artificially highlighting one part that would be much smaller in most use cases?

**Comment by gaaclarke** [comment]
Yea, this code has been production code for many years and we've got direct feedback from large clients about platform channel performance, specifically when evaluating for add-to-app scenarios where there is more communication over channels.  I think we are beyond "premature optimization" =)

The benchmark isn't absurd.  The list size is 1000 elements of various small types.  The performance regression is 13% slower.  Here's where the test payload is created: https://github.com/flutter/flutter/blob/71139099c5c4069f13dba4387b98af6158735f62/dev/benchmarks/platform_channels_benchmarks/lib/main.dart#L199

**Comment by stuartmorgan** [comment]
> we've got direct feedback from large clients about platform channel performance

Was it about Lists of 1000s of objects? I'm well aware that platform channel performance has been an issue before. I've been consistently saying that I don't see any reason to believe that the performance of *this particular case* is likely to be important, not that *no* platform channel performance matters.

> specifically when evaluating for add-to-app scenarios where there is more communication over channels.

There is a difference between "more communication" and "communication with these specific properties".

> I think we are beyond "premature optimization" =)

I have asked several questions that would help me evaluate whether that's actually true:
- "What's the real-world use case for a huge number of objects in a list over platform channels?"
- "What is the impact (either in ms or percentage of overall message time) of just this retain/release in a list with, say, 10 objects?"

I haven't seen answers to those questions, just assertions that platform channel performance matters in some general way—which, again, I am not, and have not, disputed.

> The benchmark isn't absurd. The list size is 1000 elements of various small types.

I didn't say that it was absurd, just that it doesn't seem like it would reflect typical use cases. Just saying it isn't absurd doesn't address my questions, it just dismisses them.

**Comment by gaaclarke** [comment]
> Just saying it isn't absurd doesn't address my questions, it just dismisses them.

I think looking at it that way is an unfair characterization.

Ultimately what you are asking for is analytics about real-world usage of platform channels.  Then we can know exactly what cases matter the most.  We don't have that, can't have that and even if we did have it we'd still have to generalize a solution that balances the needs of all usages.  The best we can do is extrapolate based on what users have told us.  That information is confidential.  I'm happy to discuss privately.  Xiao is another good resource if you want.

I think you are getting a bit bogged down about the maintenance being easier with ARC, but if you look at the blame for FlutterCodecs.mm the majority of it was written 6 years ago and hasn't changed in any meaningful way.  In fact, it's very sensitive to any change, more than probably any part of the engine.  So what really is the benefit of changing a file to ARC that doesn't change and has years of production testing at the cost of performance?  Even if, for arguments sake, if that performance drop is rare?

**Comment by stuartmorgan** [comment]
> Ultimately what you are asking for is analytics about real-world usage of platform channels. Then we can know exactly what cases matter the most. We don't have that, can't have that and even if we did have it we'd still have to generalize a solution that balances the needs of all usages.

I am not asking for a complete accounting for every single current and future use of platform channels, I'm asking for very specific information about what we already know.

> The best we can do is extrapolate based on what users have told us.

Agreed, which is why I'm asking what we already know about use cases based on what we've been told. Unlike your request above that I prove an impossible negative, I'm just asking whether we have actual evidence that this benchmark corresponds to a real-world use case (since it doesn't at all match up with anything I've personally seen, and I'm struggling to imagine a non-contrived use case). Either we do have such evidence, in which case this is an easy question to answer and thus resolve my concern, or we don't, in which case I continue to be skeptical about that benchmark reflecting real-world outcomes rather than being a potentially misleading microbenchmark (and in which case I'd like to know the regression for usage that looks like everything I've seen in the wild, which is lists several orders of magnitude smaller than this).

> That information is confidential. I'm happy to discuss privately. Xiao is another good resource if you want.

I can follow up privately if needed, but let me rephrase my question to avoid issues of confidentiality first: have you seen a real-world use case that looks anything like a single message containing a 1000-object list?

> So what really is the benefit of changing a file to ARC that doesn't change and has years of production testing at the cost of performance?

Having pockets of MRC code in an ARC codebase is a significant liability, because it is extremely easy for people to change it and/or review it without realizing that it's MRC:
- Leaky MRC code looks **exactly like** correct ARC code, and compiles just fine.
- It's very easy to have changes/diffs where **nothing** in the surrounding context would indicate to someone that it is MRC code (and thus needs to be reviewed as such).

When I say that it's extremely easy, I say this as someone with years of experience bouncing back and forth between ARC and non-ARC reviews (in the context of readability reviews in particular), and even when I knew to be vigilant because it was new code I still sometimes got it wrong. If we fast-forward, say, a year, at which point this could easily be the only MRC Obj-C code in the engine, that would become even more likely because nobody would be expecting to see MRC code in a review. Having the file be rarely edited is a double-edged sword because while it reduces the opportunity for mistakes, it also means people are even less likely to be on the lookout for it making them more likely to happen when those rare changes do come up.

**Comment by stuartmorgan** [comment]
> 1. We had a multi month effort to crawl back performance from the startup time of their app, where platform channels was a significant portion.  It would be an mistake to dismiss any portion of time in this code as inconsequential when we have a real world use case and a large investment in `customer:money` that proves otherwise.

Also, wasn't this the investigation that concluded that specifically optimizing platform channel encode/decode performance, even though it showed up in CPU profiles, actually had almost no measurable real-world impact on startup time? Because if so, that seems to directly support the argument that microoptimization of this particular piece of code is not so high priority that it should block ARC migration.

**Comment by gaaclarke** [comment]
> Can you point me to the part of the conversation about readValue? I only see discussion of writeValue.

You can see readValue's cost in the CPU traces.  Keep in mind these were recorded on Android and, at least the first CPU trace, is just capturing the ui thread.  The reading cost there is equal to the writing cost (it's helpful to do a pivot on StandardMessageCodec).

> This continues to conflate "real world use case of platform channel performance being important" and "real world use case of this particular piece of platform channels being important".

Right, we can not include every use case as a benchmark so it requires a bit of extrapolation.

> I'm not sure where in the thread that is, perhaps you can point me to the specific part offline. But payload size doesn't tell us anything relevant to this benchmark regression without knowing what the composition of the payload is. Adding fixed-size-per-item overhead to list decoding will have ~zero effect on a list of two 15kb objects, and a lot of effect on a list of 30,000 1-byte objects.

We don't have analytics about the actual composition of the payload.  What we do know is that they are using the StandardMessageCodec, so it is unlikely to be one item.  The idea that the contributing factor to the regression is List size is a theory.  One that was put into question when Chris extracted that function and compiled it in MRC but didn't see an improvement.  We know the channel that is generating these large payloads and we have contacts on the team that may know the payloads if we want to dig deeper.

> The specific piece that decodes per-list-item entries is O(n), yes, but unless the benchmark only tests that one exact piece of code (in which case it would definitely be a really unrealistic microbenchmark), then the overall benchmark won't scale linearly.

I don't think I understand this point.  That particular benchmark was designed for test coverage, to make sure are testing all datatypes and the size was chosen based off of choosing a size whose probability of being see in the wild is greater than 1%.  We can always create another one if we want.

---

Stepping back here's the take aways you should have from looking at the conversation with customer:money:

1) On the question of wether this regression represents the needs of any users, the probability of that should be increased in your mind since we've shown the payload size and technique are in fact used in real world applications.
1) On the question of wether the performance of platform channels matters, looking at the traces of customer:money's app startup, the timescales of what we are trying to optimize and their investment in the problem;  the probability that this regression matters should have also increased in your mind as well.

Now the question is: what is the probability that this regression will not negatively affect any users?  I suspect that should be very low given our limited data.  So hopefully we can dismiss the notion that this regression wouldn't affect any user or that this microbenchmark is unrepresentative on account of that being unlikely.

So, now the calculus becomes simply: "Is the cost of maintaining one MRC file that rarely changes and has 6 years of testing greater than the cost of negatively affecting users performance?"

**Comment by stuartmorgan** [comment]
> Now the question is: what is the probability that this regression will not negatively affect any users?

That is not the only question. "How many users will be affected" and "how much will users be affected" are also very important questions.

> So, now the calculus becomes simply: "Is the cost of maintaining one MRC file that rarely changes and has 6 years of testing greater than the cost of negatively affecting users performance?"

There are several implicit value judgements baked into that "simple" distillation of the problem. I could just as accurately say that the calculus is "Is the cost of permanently maintaining a mixed MRC/ARC codebase, thus increasing the probability of future crash and leak bugs, greater than the cost of regressing a benchmark whose direct applicability to real-world performance has not been clearly established?"

My opinion is that we should move forward with the ARC migration without a carve-out for the code covered by this benchmark, giving ourselves a solid foundation for future engine work. If in the future we have further evidence that platform channel encoding/decoding in particular is a sufficiently high priority for revisiting for optimization, we can investigate the best way to optimize it (which, per comments in the related doc, would probably *not* be via MRC ObjC, but some other solution).

**Comment by jmagman** [comment]
> My opinion is that we should move forward with the ARC migration without a carve-out for the code covered by this benchmark, giving ourselves a solid foundation for future engine work. If in the future we have further evidence that platform channel encoding/decoding in particular is a sufficiently high priority for revisiting for optimization, we can investigate the best way to optimize it (which, per comments in the related doc, would probably not be via MRC ObjC, but some other solution).

I completely agree.

**Comment by gaaclarke** [comment]
> Also, wasn't this the investigation that concluded that specifically optimizing platform channel encode/decode performance, even though it had showed up in CPU profiles, actually had almost no measurable real-world impact on startup time? Because if so, that seems to directly support the argument that microoptimization of this particular piece of code is not so high priority that it should block ARC migration.

My understanding from working on that team is that there were no easy solutions and that multiple fronts had to be addressed to get the total goal.  In the buganizer issue linked above you see a customer mentioning that it's hard to measure real world performance because it's very noisy.  They had to look at P95 numbers to detect a 2% decrease in startup time as a result of avoiding platform channel payload copies.  While that is hard to detect, a 2% decrease is actually not insignificant as you'll see below.

Here's another example which is tied to the performance of the encoder: We noticed that `customer:money` was bottlenecked on the platform thread, which includes the code in question.  In order to address this for `customer:money` we created the background platform channels which allows the encoding (and execution of the handlers) to happen on a background thread.  Migrating one plugin over resulted in a 3.5% (47ms) decrease in startup time (b/237035671).  I can't remember how much time we were trying to shave off, but let's say it was 2 seconds, that was 1/40th of our goal.  We had multiple similar small gains and a few bigger ones, but that's how we reached the goal.

Note that while background platform channels provides a workaround for this regression, it isn't always possible to use them.

So on the question of wether encoding changes were measurable, what we do know is that the variance of real world benchmarks can hide significant changes.  That's why we must consider the microbenchmarks, real-world benchmarks, and profiles when drawing conclusions.  Not one single measurement paints the whole picture.  If you want to link to me privately some results you are looking at I can better respond to them.

**Review by stuartmorgan — APPROVED** [review]
> In the buganizer issue linked above you see a customer mentioning that it's hard to measure real world performance because it's very noisy. They had to look at P95 numbers to detect a 2% decrease in startup time as a result of avoiding platform channel payload copies.

That is not what it says though. It says that with a benchmark resolution of 2% at P95 the benchmark showed **no improvement**.

> While that is hard to detect, a 2% decrease is actually not insignificant as you'll see below.

"2%" and "0% +/-2%" are not the same thing.

> Here's another example which is tied to the performance of the encoder: We noticed that customer:money was bottlenecked on the platform thread

It also has a lot to do with real-world thread contention, which this benchmark does not measure. The fact that encoder/decoder benchmark performance and real-world end-to-end performance are not at all the same thing is the core of my position.

> we created the background platform channels which allows the encoding (and execution of the handlers) to happen on a background thread.

You are continuing to argue against a straw man that I have been very clear is not my position. I have, at no point, suggested that end-to-end, real-world platform channel performance does not matter.

Nobody in this review has advocated for the removal of background platform channel support. To the extent that the background channel data is relevant at all to the PR at hand, I believe it supports my position: in the real-world example you are drawing all of this data from the outcome was that
- optimizing serialization had *no measurable impact* on startup time, while
- moving the end-to-end process to another thread had a significant effect on startup time.

So the evidence is that encoder/decoder performance *by itself*, which is what this PR discussion is about, was actually not a significant driver of real-world perf outcomes based on the data we have, and that other parts of the system play a significantly larger role.

I don't think continued discussion involving me will be productive here, as we are both now repeating things we've already said. I am comfortable with the tradeoff of proceeding here for the reasons I've articulated at length above (and summarized at the end of [this comment](https://github.com/flutter/engine/pull/37883#issuecomment-1339953658)), so I'm approving this from my standpoint.

@cyanglaz given the disagreement here, please don't land based on my approval alone. At least @jmagman and @zanderso should approve (or not) before proceeding, and potentially Ian as well if this needs to be escalated.

**Review by gaaclarke — COMMENTED** [review]
We converted some retains to copies which may be contributing to some of the regression.  The original PR didn't raise the question about wether copying the data is the correct course here.

**Review by gaaclarke — CHANGES_REQUESTED** [review]
Request change not because of the larger issue discussed, but I want to make sure we don't land this without explanation why some retains were changed to copies =)

File: shell/platform/darwin/common/framework/Source/FlutterStandardCodec.mm [hunk 1]
```diff
@@ -348,16 +328,11 @@ @implementation FlutterStandardReader {
 - (instancetype)initWithData:(NSData*)data {
   self = [super init];
   NSAssert(self, @"Super init cannot be nil");
-  _data = [data retain];
+  _data = [data copy];
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/platform/darwin/common/framework/Source/FlutterStandardCodec.mm:331]: Are you sure NSData is copy-on-write?  Data in Swift is, I don't see it mentioned in the NSData documentation.

[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/platform/darwin/common/framework/Source/FlutterStandardCodec.mm:331]: Copy-on-write may also not mean that creating the copy is free.  If they are using VM pages to create copy-on-write segments it actually would require syscalls to the kernel to make that happen.  I think it's worth trying the benchmark with the old memory semantics.

Here's the output to a tiny app that runs [NSData copy] on 100MB of NSData:
Screen Shot 2022-12-07 at 2 48 17 PM

I'm guessing the fact that it isn't considered heap allocation points to using copy-on-write mmap like I speculated above.

**Comment by gaaclarke** [comment]
Okay, I tested moving the code to CoreFoundation.  I had forgotten about `CFAutorelease` so we can probably duplicate MRC that way.  The simple test I did was faster with CoreFoundation, it wasn't quite to the levels we were seeing with MRC but the assembly looks close and I think we can probably duplicate it.

```objc
- (nullable id)readValue {
  return (__bridge id)ReadValue(self);
}

static CFTypeRef ReadValue(FlutterStandardReader* codec) {
  return ReadValueOfType(codec, [codec readByte]);
}

- (nullable id)readValueOfType:(UInt8)type {
  return (__bridge id)ReadValueOfType(self, type);
}

static CFTypeRef ReadValueOfType(FlutterStandardReader* codec, UInt8 type) {
  FlutterStandardField field = (FlutterStandardField)type;
  switch (field) {
    case FlutterStandardFieldNil:
      return nil;
    case FlutterStandardFieldTrue:
      return (__bridge CFBooleanRef) @YES;
    case FlutterStandardFieldFalse:
      return (__bridge CFBooleanRef) @NO;
    case FlutterStandardFieldInt32: {
      SInt32 value;
      [codec readBytes:&value length:4];
      return (__bridge CFNumberRef) @(value);
    }
    case FlutterStandardFieldInt64: {
      SInt64 value;
      [codec readBytes:&value length:8];
      return (__bridge CFNumberRef) @(value);
    }
    case FlutterStandardFieldFloat64: {
      Float64 value;
      [codec readAlignment:8];
      [codec readBytes:&value length:8];
      return (__bridge CFNumberRef)[NSNumber numberWithDouble:value];
    }
    case FlutterStandardFieldIntHex:
    case FlutterStandardFieldString:
      return (__bridge CFStringRef)[codec readUTF8];
    case FlutterStandardFieldUInt8Data:
    case FlutterStandardFieldInt32Data:
    case FlutterStandardFieldInt64Data:
    case FlutterStandardFieldFloat32Data:
    case FlutterStandardFieldFloat64Data:
      return (__bridge CFTypeRef)[codec readTypedDataOfType:FlutterStandardDataTypeForField(field)];
    case FlutterStandardFieldList: {
      UInt32 length = [codec readSize];
      CFMutableArrayRef array = CFArrayCreateMutable(kCFAllocatorDefault, length, &kCFTypeArrayCallBacks);
      for (UInt32 i = 0; i < length; i++) {
        CFTypeRef value = ReadValue(codec);
        CFArrayAppendValue(array, (value == nil ? kCFNull : value));
      }
      return CFAutorelease(array);
    }
    case FlutterStandardFieldMap: {
      UInt32 size = [codec readSize];
      CFMutableDictionaryRef dict =
          CFDictionaryCreateMutable(kCFAllocatorDefault, size, &kCFTypeDictionaryKeyCallBacks,
                                    &kCFTypeDictionaryValueCallBacks);
      for (UInt32 i = 0; i < size; i++) {
        CFTypeRef key = ReadValue(codec);
        CFTypeRef val = ReadValue(codec);
        CFDictionaryAddValue(dict, (key == nil ? kCFNull : key), (val == nil ? kCFNull : val));
      }
      return CFAutorelease(dict);
    }
    default:
      NSCAssert(NO, @"Corrupted standard message");
  }
}
```

Here is the profile of the test (it's 4.72s above):
Screen Shot 2022-12-07 at 9 23 32 PM

Here's the assembly output of the `FlutterStandardFieldInt32` branch for ARC and MRC with this change:
Screen Shot 2022-12-07 at 9 26 13 PM

I'm comfortable with this change if we migrate the rest of the code to CoreFoundation.  I can finish up the conversion tomorrow quick now that we have a nice proof of concept.  I have to double check that this doesn't choke on `FlutterStandardTypedData`.

**Review by gaaclarke — CHANGES_REQUESTED** [review]
The copy I mentioned looked fine.  If we just migrate the codec to CoreFoundation this sounds good to me.

**Comment by cyanglaz** [comment]
After investigation, detailed in this [doc](https://docs.google.com/document/d/1D90pSzlhFjreJwrsAXi9fL-P9skttl9wnS3EH1zA3iQ/edit#heading=h.endy2s5qcg5b) We have determined that migrating to static C function and CoreFoundation is not trivial. Discussed with @gaaclarke offline and we think we should move on and accept the regression for now and come back revisit the performance of platform channel later.

**Review by gaaclarke — APPROVED** [review]
Thanks @cyanglaz for looking into this.  I strongly believe this migration to ARC should be limited to logic that does not affect benchmarks on the grounds that customer needs trump developer comfort, especially when the gravity of the benchmark is weighed against the utility of having this one file be ARC.  But assuming the team has considered that and does agree with the tradeoff; this code is correct and a fair effort was made to improve it.

**Review by zanderso — APPROVED** [review]

**Review by jmagman — APPROVED** [review]
